src/main/java/ch/njol/skript/lang/function/Parameter.java

```diff
@@ -58,18 +61,72 @@ public final class Parameter<T> {
 	 */
 	final boolean keyed;
 
+	@Deprecated(since = "INSERT VERSION", forRemoval = true)
 	public Parameter(String name, ClassInfo<T> type, boolean single, @Nullable Expression<? extends T> def) {
 		this(name, type, single, def, false);
 	}
 
+	@Deprecated(since = "INSERT VERSION", forRemoval = true)
 	public Parameter(String name, ClassInfo<T> type, boolean single, @Nullable Expression<? extends T> def, boolean keyed) {
 		this.name = name;
 		this.type = type;
 		this.def = def;
 		this.single = single;
+		this.optional = def != null;
 		this.keyed = keyed;
 	}
 
+	/**
+	 * Constructs a new parameter.
+	 *
+	 * @param name The name.
+	 * @param type The type of the parameter.
+	 * @param modifiers The modifiers of this parameter.
+	 */
+	public Parameter(@NotNull String name, @NotNull Class<T> type, Modifier... modifiers) {
+		Preconditions.checkNotNull(name, "name cannot be null");
+		Preconditions.checkNotNull(type, "type cannot be null");
+
+		Set<Modifier> mods = Set.of(modifiers);
+
+		this.name = name;
+		this.type = DefaultFunction.getClassInfo(type);
+		this.def = null;
+		this.single = !type.isArray();
+		this.optional = mods.contains(Modifier.OPTIONAL);
+		this.keyed = mods.contains(Modifier.KEYED);
+	}
+
+	/**
+	 * Represents a modifier that can be applied to a parameter
+	 * when constructing one using {@link Builder#parameter(String, Class, Modifier[])}}.
+	 */
+	public interface Modifier {
+
```

Comment:
i would have a method like. That should also let you replace the ModifierImpl class
```suggestion

	/**
	 * @return A new Modifier instance to be used as a custom flag.
	 */
	static Modifier of() {
		return new Modifier() { };
	}

```